alephium/contracts/token_bridge/attest_token_handler.ral

```diff
@@ -0,0 +1,37 @@
+Contract AttestTokenHandler(
+  governanceContractId: ByteVec,
+  localChainId: U256,
+  localTokenBridgeId: ByteVec,
+  remoteChainId: U256,
+  remoteTokenBridgeId: ByteVec,
+  mut receivedSequence: U256
+) extends Constants(), TokenBridgeConstants() {
+  @using(preapprovedAssets = true)
+  pub fn handleAttestToken(
+    vaa: ByteVec,
+    payer: Address,
+    createContractAlphAmount: U256
+  ) -> () {
+    let governance = Governance(governanceContractId)
+    let (emitterChainId, targetChainId, emitterAddress, msgSequence, payload) = governance.parseAndVerifyVAA(vaa)
+    assert!(emitterChainId == remoteChainId, ErrorCodes.InvalidEmitChainId)
+    assert!(emitterAddress == remoteTokenBridgeId, ErrorCodes.InvalidEmitAddress)
+    assert!(targetChainId == localChainId || targetChainId == 0, ErrorCodes.InvalidTargetChainId)
+    assert!(msgSequence >= receivedSequence, ErrorCodes.InvalidSequence)
```

Comment:
1. No, the `AttestToken` message is sent by the contract on the chain, and the sequence will increase by 1 every time a message is sent. For example, ff the user wants to attest a token on ETH to Alephium, the user needs to call the ETH contract, and it will publish the `AttestToken` message, but If the user did not complete the attestation with the `AttestToken` message, then the message may become invalid later, and user need to re-do attestation.
2. No, the contract only needs to ensure that every time an `AttestToken` message is sent, the sequence will be incremented by one.

Reply:
Thanks for the explanation :+1:

Reply:
I understand why we use sequence to prevent replay attack. I guess compared to using hash, the cons is that we put higher requirement on the correctness of the smart contract on the other chain, since it now has the power to break our handler if there is a bug on sequencing on the their side, whereas with the hash approach, that is not possible.

Reply:
Yes, for other chain implementations, even if the sequence is the same, the verification can be successful as long as the message hash is different, but the sequence cannot be the same for us.

Reply:
I think sequence should be unique by design and by code:
1. The sequence is updated when publishing message, so every new message (bridge action) would have different sequence number.
2. Sequence seems designed to be unique:
```
        // Nonce of the VAA, must to be set to random bytes. Nonces
	// prevent collisions where one emitter publishes identical
	// messages within one block (= timestamp).
	//
	// It is not suitable as a global identifier -
	// use the (chain, emitter, sequence) tuple instead.
	Nonce uint32 // <-- NEW
```
3. A lot of wormhole's code relies on sequence number. If sequence cannot be used as UUID, then it will be a problem for them too.